### PR TITLE
cache build, shard and split actions

### LINKED ISSUE #20072: improve CI run time + CI total time
### Describe the Improvement

1. cache typescript builds so they are only done build 1 time per commit. (decrease CI total time)
2. shard tests (increase CI total time but decrease CI run time).

end goal make CI time faster without clogging up pipeline more the currently is.

note that we can only run 60 runners at the same time. this means if if CI total gets to high. it could lead to clogging up github actions so that you need to wait until runners become free of there previous task
```[tasklist]
### TODO
- [x] cache typescript builds
- [x] shard tests
```

### PR DESCRIPTION
**Pull request by Boegie19**
## Scope

What's changed:

- split build and install of from prepare actions.
- cache build
- add build step before test step (to only but once instead of for every shard)
- shard 3 times all bot blackbox-pr and blackbox-main
- removed build step for with build false. (since now build is split of from prepare)

## Potential Risks / Drawbacks
- You can only run 2 merges in parallel before the merge starts backing up (still the merge time should be more then halved before this was 6 before it got backed up
## Review Notes / Questions
- This change should be tested and I think 
- If you don't want build and install actions to be split form prepare I could undo that change. (Need it how ever to confirm I did not break the build step is not broken)

---

Fixes #20072

### CODE REVIEW

File: .github/workflows/blackbox-pr.yml [hunk 1]
```diff
@@ -21,19 +21,45 @@ env:
...
+        uses: ./.github/actions/build
+      
+
       - name: Start services
         run:
           docker compose -f tests/blackbox/docker-compose.yml up auth-saml redis minio minio-mc -d --quiet-pull --wait
 
       - name: Run tests
-        run: TEST_DB=sqlite3 pnpm run test:blackbox
+        run: TEST_DB=sqlite3 pnpm run test:blackbox --shard=${{ matrix.shard }}
```
Comment (paescuj) [inline .github/workflows/blackbox-pr.yml:65]:
```suggestion
        run: TEST_DB=sqlite3 pnpm run test:blackbox -- --shard=${{ matrix.shard }}
```

Comment (Boegie19) [reply .github/workflows/blackbox-pr.yml]: just realized the same thing.

**Comment by Boegie19** [comment]
btw I now understand why tests are randomly failing to much you guys run a container in a container.

**Comment by paescuj** [comment]
Very interesting @Boegie19!
I'm curious about the `shard` option, basically a great match for our blackbox test. Not entirely sure whether that works with our setup, because it's quite special (custom sequencer, etc.) - but let's see 👀 😄 

Please do not spend all too much time on this, though, at least not without discussing with us beforehand. We're planning to do some major restructuring of the workflows resp. build / tests in the coming weeks, so your work might conflict.

**Comment by paescuj** [comment]
> btw I now understand why tests are randomly failing to much you guys run a container in a container.

Haha it's more because the tests are (too) CPU intensive & time sensitive

**Comment by Boegie19** [comment]
@paescuj shard split the work between multiple vms aka every vm gets a shard 1/3 and it will do the first 1/3 of the tests. This only has some issues. as in some work now needs to be done 2 times.

**Comment by Boegie19** [comment]
> > btw I now understand why tests are randomly failing to much you guys run a container in a container.
> 
> Haha it's more because the tests are (too) CPU intensive & time sensitive

I think that using the github build in way to do it.

services:
      mysql:
        image: bitnami/mysql:5.7 
        
 would maby make it better not sure do.

**Comment by paescuj** [comment]
Closing for now as this approach doesn't work! Thanks anyway @Boegie19  for the exciting experiment! 😃

**Pull request closed without merging**